File: Electrical/Tests/spiTest/masterTest.cpp
#include <wiringPi.h>
#include <wiringPiSPI.h>

int main(void)
{
	const int stmPin = 4;
	unsigned char dataOn[1];
	unsigned char dataOff[1];

	// init wiringPi
	wiringPiSetup();
	// init SPI at channel 0 at 25 mHz clock speed
	wiringPiSPISetup(0, 25000000);
	
	// set the stm32 slave select pin as output and set high
	pinMode(stmPin, OUTPUT);
	digitalWrite(stmPin, HIGH);

	// turn led on stm32 on and off 10 times
	for (int i = 0; i < 10; i++)
	{
		// data values are set every loop because the wiringPiSPI funciton writes the data and then overwrites it with what is read from the stm32
		*dataOn = 0b11111111;
		*dataOff = 0b00000000;

		// pull low to select stm32 and wait so that the stm32 is listening
		digitalWrite(stmPin, LOW);
		// wiringPi also has a delay function like arduino
		delay(10);

		// send the data on channel zero and read 1 byte (reading 0 doesnt work)
		wiringPiSPIDataRW(0, dataOn, 1);
		digitalWrite(stmPin, HIGH);
		delay(50);

		digitalWrite(stmPin, LOW);
		delay(10);

		wiringPiSPIDataRW(0, dataOff, 1);
		digitalWrite(stmPin, HIGH);
		delay(50);
	}
	return 0;
}

File: Electrical/Tests/spiTest/masterTest.py
# This code uses spi on the raspberry pi to send data to the stm32 running the slaveTest code
import RPi.GPIO as GPIO # To use gpio on the pi
import spidev # to use spi on the pi
import time

# set up spi with chip enable 0 (we will do chip enable manually) on spi channel 0
spi = spidev.SpiDev()
spi.open(0, 0)

# set pinout to BCM pinout (what you see when you look up raspberry pi pinout)
GPIO.setmode(GPIO.BCM)

# set up pin 4 for slave select as output (very similar to arduino) and pull high
GPIO.setup(4, GPIO.OUT)
GPIO.output(4, GPIO.HIGH)

# create variables that hold the messages to be sent (msg1 will turn the light on the stm32 on and msg will turn it off)
msg1 = 0b10101
msg2 = 0b01010

# This loop will turn the light on the stm32 on and off ten times
for i in range(10):
    # select stm32 by pulling pin 4 low, wait to make sure stm32 is listening and send data and then pull high to deselect
    GPIO.output(4, GPIO.LOW)
    time.sleep(0.01)
    spi.xfer([msg1]) # xfer requires a list to send the data
    GPIO.output(4, GPIO.HIGH)

    time.sleep(0.05)

    # do the same as before but send 0
    GPIO.output(4, GPIO.LOW)
    time.sleep(0.01)
    spi.xfer([msg2])
    GPIO.output(4, GPIO.HIGH)

    time.sleep(0.05)

# clean up GPIO
GPIO.cleanup()
